include/pybind11/detail/type_caster_base.h

```diff
@@ -514,7 +515,14 @@ PYBIND11_NOINLINE std::string error_string() {
             errorString += "  " + handle(f_code->co_filename).cast<std::string>() + "("
                            + std::to_string(lineno)
                            + "): " + handle(f_code->co_name).cast<std::string>() + "\n";
-            frame = frame->f_back;
+#    if PY_VERSION_HEX >= 0x030900B1
+            auto *b_frame = PyFrame_GetBack(frame);
+#    else
+            auto *b_frame = frame->f_back;
+            Py_XINCREF(b_frame);
```

Comment:
This the leak?

Reply:
It's on 3.9 so it seems the issue is `PyFrame_GetBack(frame)`? It will be decremented on the next iteration. If it's null, it doesn't need to be decremented. I am not seeing the issue with the control flow.

Reply:
I think it's the fact that the b_frame doesn't get DECREF'd at the end of the loop. The frame does, but not the b_frame.